### PR TITLE
Quantum channels

### LINKED ISSUE #1193: methods for developing quantum channels
Go with pr #1191 

Also @chriseclectic can you make an issue for all the other future PR that you list there.

### PR DESCRIPTION
**Pull request by chriseclectic**
Closes #1193 

### Summary

Adds quantum channel package to `quantum_info`. This package contains classes for representing quantum channels (CPTP maps, and also general Non-CPTP operator maps) and is essentially an implementation of the paper: 

*Tensor networks and graphical calculus for open quantum systems*
Wood, Biamonte, Cory, Quant. Inf. Comp. 15, 0579-0811 (2015)
Open access: [arXiv: arXiv:1111.6950v3](https://arxiv.org/abs/1111.6950)

### Details and comments

* Contains classes for channel representations: 
  * `Choi`: Choi-Jamiolkowski process matrix
  * `SuperOp`: Linear superoperator
  * `Chi`: Pauli-basis Chi matrix
  * `PTM`: Pauli transfer-matrix superoperator
  * `Kraus`:  Kraus / operator-sum operators
  * `Stinespring`: Stinespring / system-environment insometry
  * `UnitaryChannel`: single-matrix Kraus representation (not generic operator map like other reps, and not exposed to user by default)

* The classes can be converted into each other using their init methods eg: `SuperOp(Kraus([k1, k2]))`
* All classes contain methods for
  * `is_cptp`: check if map is completely positive trace preserving
  * `evolve`: evolve a quantum state by the channel
  * `adjoint`: compute to the adjoint channel
  * `conjugate`: compute the conjugate channel
  * `transpose`: compute the transposed channel
  * `compose`: compose with another channel B(A(rho))
  * `tensor`: tensor product with another channel A \otimes B
  * `expand`: expand with another channel B \otimes A (reverse order tensor product)
  * `power`: compose channel with itself n times
  * `add`: linear addition of two channels
  * `subtract`: linear subtraction of two channels
  * `multiply`: scalar multiplication of a channel
* Operator overloads:
  * `+,-,*` for add, subtract, multiply
  * `@` for compose
  * `^` for tensor

Closes #1193 

### Tests

- [x] Tests for `Choi` class
- [x] Tests for `SuperOp` class
- [x] Tests for `Chi` class
- [x] Tests for `PTM` class
- [x] Tests for `Stinespring` class
- [x] Tests for `Kraus` class
- [x] Tests for `UnitaryChannel` helper class
- [x] Equivalence tests for evolution
- [x] Equivalence tests for tensor, expand
- [x] Equivalence tests for compose
- [x] Equivalence tests for conjugate, transpose, adjoint
- [x] Equivalence tests for add, subtract, multiply

### Future PR:

- Integrate with Ericks `Unitary` class in PR #1961 
- Integrate with Qiskit Aer noise model

### CODE REVIEW

**Comment by ajavadia** [comment]
I think this closes #1193

**Comment by chriseclectic** [comment]
@jaygambetta @ajavadia this is finished and ready for review

File: test/python/quantum_info/channel/test_chi.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+
+    def test_compose(self):
+        """Test compose method."""
+        # Random input test state
+        rho = self.rand_rho(2)
+
+        # UnitaryChannel evolution
+        chan1 = Chi(self.chiX)
+        chan2 = Chi(self.chiY)
+        chan = chan1.compose(chan2)
```
Comment (jaygambetta) [inline test/python/quantum_info/channel/test_chi.py]: I not sure I really like this order

**Comment by jaygambetta** [comment]
also chris can we make an issue to make the PTM change order depending on the order a user gives it. Ie I should be able to put in a Pauli group.

**Comment by chriseclectic** [comment]
Changed order of tensor product and removed the `front` kwarg to swap order. The reverse order tensor product is now done with a seperate method called `expand`

File: test/python/quantum_info/channel/test_chi.py (hunk 1 above)
Comment (chriseclectic) [reply test/python/quantum_info/channel/test_chi.py]: We need this order to be consistent with the rest of the circuit gate type objects.

File: test/python/quantum_info/channel/base.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        return [self.rand_matrix(output_dim, input_dim) for _ in range(n)]
+
+    def assertAllClose(self,
+                       obj1,
+                       obj2,
+                       rtol=1e-5,
+                       atol=1e-6,
+                       equal_nan=False,
+                       msg=None):
+        """Assert to objects are equal using Numpy.allclose."""
```
Comment (ajavadia) [inline test/python/quantum_info/channel/base.py]:
```suggestion
        """Assert two objects are equal using Numpy.allclose."""
```

**Comment by ajavadia** [comment]
isn't there a bunch of duplication between the UnitaryChannel here and the one that derives from Gate in #1774 ?

**Comment by chriseclectic** [comment]
@ajavadia yeah it  can be removed once Ericks is finished, but its needed here internally for tests.

File: qiskit/quantum_info/operators/predicates.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        atol = ATOL_DEFAULT
+    if rtol is None:
+        rtol = RTOL_DEFAULT
+    mat = np.array(op)
+    if mat.ndim != 2:
+        return False
+    return np.allclose(mat, np.conj(mat.T), rtol=rtol, atol=atol)
+
+
+def is_positive_semidefinite_matrix(op, rtol=RTOL_DEFAULT, atol=ATOL_DEFAULT):
```
Comment (ajavadia) [inline qiskit/quantum_info/operators/predicates.py]: nit: can you make the arg of all these predicate functions `mat`? currently some are `op` and some are `mat`.

**Comment by ajavadia** [comment]
Nice work, especially writing all those unit tests.

This generally looks good to me. I have one concern however: what is the actual use case of the `inplace` args? At first I liked the idea, but as I was working on #1816 I realized this is not as easy to do, since:
- we don't want to allow modifying the fundamental objects (say H gate). 
- it becomes a bit ambiguous in hierarchical setting (modifying circuit A in place, when it contains circuit B; does B get modified?). 
- it complicates the code a lot, as you have `if inplace:` checks all over.
- we open up circuit/channel objects to being modified, which may potentially cause hard-to-track bugs later.

So what do we lose if we just said every operation returns a new circuit or channel? I'm mainly talking about the end user scenario, not internal implementation (cc'ing @kdk).

**Pull request merged**